File: src/lib/scala/Slate.Api/src/main/scala/slate/core/apis/svcs/ApiEntityWithSupport.scala
/**
<slate_header>
  author: Kishore Reddy
  url: https://github.com/kishorereddy/scala-slate
  copyright: 2015 Kishore Reddy
  license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  desc: a scala micro-framework
  usage: Please refer to license on github for more info.
</slate_header>
  */

package slate.core.apis.svcs

import scala.reflect.runtime.universe.{Type}
import slate.common.encrypt.{EncryptSupportIn}
import slate.common.results.ResultSupportIn
import slate.core.common.{AppContext, AppContextSupport}
import slate.entities.core.{Entity}
import slate.common.i18n.{I18nSupportIn}
import slate.common.logging.{LogSupportIn}
import slate.core.apis.{ApiBaseEntity}



class ApiEntityWithSupport[T >: Null <: Entity, TSvc >: Null]
(
  context:AppContext, tpe:Type
)
  extends ApiBaseEntity[T](context, tpe)
  with EncryptSupportIn
  with LogSupportIn
  with I18nSupportIn
  with ResultSupportIn
  with AppContextSupport
{
  def service: TSvc = _service.asInstanceOf[TSvc]


  override protected def log() = Option(context.log)
  override protected def enc() = context.enc
  override protected def res() = context.res
}

File: src/lib/kotlin/slatekit-apis/src/main/kotlin/slatekit/apis/svcs/ApiEntityWithSupport.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.apis.svcs

import slatekit.apis.ApiBaseEntity
import slatekit.common.encrypt.EncryptSupport
import slatekit.common.encrypt.Encryptor
import slatekit.common.log.LogSupport
import slatekit.common.log.LoggerBase
import slatekit.core.common.AppContext
import slatekit.entities.core.Entity
import slatekit.entities.core.EntityService
import kotlin.reflect.KClass


open class ApiEntityWithSupport<T, TSvc>(
        context: AppContext, val type: KClass<*>
) : ApiBaseEntity<T>(context, type), EncryptSupport, LogSupport where T : Entity, TSvc : EntityService<T> {
    override val logger: LoggerBase? get() = context.log
    override val encryptor: Encryptor? get() = context.enc


    @Suppress("UNCHECKED_CAST")
    fun service(): TSvc = context.ent.getSvc<T>(type) as TSvc
}
